tests/core/vm/test_vm_state.py

```diff
@@ -47,3 +57,163 @@ def test_state_db(state):  # noqa: F811
     with state.state_db(read_only=True) as state_db:
         with pytest.raises(TypeError):
             state_db.set_balance(address, 0)
+
+
+def test_apply_transaction(chain_without_block_validation):  # noqa: F811
+    chain = chain_without_block_validation  # noqa: F811
+
+    # Don't change these variables
+    vm = chain.get_vm()
+    chaindb = copy.deepcopy(vm.chaindb)
+    block0 = copy.deepcopy(vm.block)
+    prev_block_hash = chain.get_canonical_block_by_number(0).hash
+    initial_state_root = vm.state.block_header.state_root
+
+    # (1) Get VM.apply_transaction(transaction) result for assertion
+    # The first transaction
+    chain1 = copy.deepcopy(chain)
+    vm_example = chain1.get_vm()
+    vm_example._is_stateless = False  # Only for testing
+    recipient1 = decode_hex('0x1111111111111111111111111111111111111111')
+    amount = 100
+    from_ = chain.funded_address
+    tx1 = new_transaction(
+        vm_example,
+        from_,
+        recipient1,
+        amount,
+        private_key=chain.funded_address_private_key,
+    )
+    computation, result_block = vm_example.apply_transaction(tx1)
+
+    # The second transaction
+    recipient2 = decode_hex('0x2222222222222222222222222222222222222222')
+    tx2 = new_transaction(
+        vm_example,
+        from_,
+        recipient2,
+        amount,
+        private_key=chain.funded_address_private_key,
+    )
+    computation, result_block = vm_example.apply_transaction(tx2)
+    assert len(result_block.transactions) == 2
+
+    # (2) Test VMState.apply_transaction(...)
+    # Use FrontierVMState to apply transaction
+    chaindb1 = copy.deepcopy(chaindb)
+    block1 = copy.deepcopy(block0)
+    block_header1 = block1.header
+    prev_headers = vm.get_prev_headers(
+        last_block_hash=prev_block_hash,
+        db=vm.chaindb,
+    )
+    vm_state1 = FrontierVMState(
+        chaindb=chaindb1,
+        block_header=block_header1,
+        prev_headers=prev_headers,
+        receipts=[],
+    )
+    parent_header = copy.deepcopy(prev_headers[0])
+
+    computation, block, _ = FrontierVMState.apply_transaction(
+        vm_state1,
```

Comment:
This is what I meant; it doesn't feel right to call a classmethod passing an instance of that same class as its first argument. I guess that's because we wanted to make `apply_transaction()` a pure function, but then maybe we should make it into a standalone function as well, to avoid this weirdness when calling it?

Reply:
True! Actually, I think after the discussion in https://github.com/ethereum/py-evm/pull/247#issuecomment-356301856 and keeping VMState mutable, we can change them back to regular instance method `def apply_transaction(self, transaction, block, is_stateless=True)`.

Reply:
True!
Actually, I think after the discussion in https://github.com/ethereum/py-evm/pull/247#issuecomment-356301856 and keeping VMState mutable, another way we can do is to change them back to regular instance method `def apply_transaction(self, transaction, block, is_stateless=True)`. And so do `VMState.execute_transaction(...)` and `VMState.make_receipt(...)` methods.

@pipermerriam what do you think?

Reply:
Yes, going back to regular instance methods seems to make the most sense.